pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -0,0 +1,483 @@
+        @Getter
+        private long addedTime;
+
+    }
+
+    /***
+     * The context for {@link ManagedLedger#asyncAddEntry(byte[], AsyncCallbacks.AddEntryCallback, Object)}, Holds the
+     * data array written in batches and callback parameters that need to be executed after batched write is complete.
+     */
+    public static class FlushContext{
+
+        private static final Recycler<FlushContext> FLUSH_CONTEXT_RECYCLER = new Recycler<FlushContext>() {
+            @Override
+            protected FlushContext newObject(Handle<FlushContext> handle) {
+                return new FlushContext(handle);
+            }
+        };
+
+        private final Recycler.Handle<FlushContext> handle;
+
+        /** Callback parameters for current batch. **/
+        private final ArrayList<AsyncAddArgs> asyncAddArgsList;
+
+        public FlushContext(Recycler.Handle<FlushContext> handle){
+            this.handle = handle;
+            this.asyncAddArgsList = new ArrayList<>(8);
+        }
+
+        public static FlushContext newInstance(){
+            return FLUSH_CONTEXT_RECYCLER.get();
+        }
+
+        public static FlushContext newInstance(int asyncAddArgsListInitialCapacity){
+            FlushContext flushContext = FLUSH_CONTEXT_RECYCLER.get();
+            flushContext.asyncAddArgsList.ensureCapacity(asyncAddArgsListInitialCapacity);
+            return flushContext;
+        }
+
+        public void recycle(){
+            this.asyncAddArgsList.clear();
+            this.handle.recycle(this);
+        }
+    }
+
+    /***
+     * The difference with {@link PositionImpl} is that there are two more parameters:
+     * {@link #batchSize}, {@link #batchIndex}.
+     */
+    public static class TxnBatchedPositionImpl extends PositionImpl {
+
+        /** The data length of current batch. **/
+        @Getter
+        private final int batchSize;
+
+        /** The position of current batch. **/
+        @Getter
+        private final int batchIndex;
+
+        public TxnBatchedPositionImpl(Position position, int batchSize, int batchIndex, long[] ackSet){
+            super(position.getLedgerId(), position.getEntryId(), ackSet);
+            this.batchIndex = batchIndex;
+            this.batchSize = batchSize;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof TxnBatchedPositionImpl other) {
+                return super.equals(o) && batchSize == other.batchSize && batchIndex == other.batchIndex;
+            }
+            return false;
+
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(super.hashCode(), batchSize, batchIndex);
+        }
+    }
+
+    interface AddDataCallback {
+
+        void addComplete(Position position, Object context);
+
+        void addFailed(ManagedLedgerException exception, Object ctx);
+    }
+
+    public enum State{
+        RUNNING,
```

Comment:
I suggest using OPEN. The `RUNNING` state sounds a little bit confusing, it can easily be interpreted as the writer are processing some write operations for now.

Reply:
Fixed.